src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +490,374 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+            // 3. (Later) the values in the map can be read to avoid providing new values for equivalent parameters.
+            if (_state._method is not null)
+            {
+                foreach (var previousParameter in _state._method.Parameters)
+                {
+                    if (ExpansionSession.GetFieldValue(previousParameter.Name, out var previousValue) == VSConstants.S_OK)
+                    {
+                        _state._arguments = _state._arguments.SetItem(previousParameter.Name, previousValue);
+                    }
+                }
+            }
+
+            // We need to replace the portion of the existing Full Method Call snippet which appears inside parentheses.
+            // This span starts at the beginning of the first snippet field, and ends at the end of the last snippet
+            // field. Methods with no arguments still have an empty "placeholder" snippet field representing the initial
+            // caret position when the snippet is created.
+            var textSpan = new VsTextSpan[1];
+            if (ExpansionSession is null || ExpansionSession.GetSnippetSpan(textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            var firstField = _state._method?.Parameters.FirstOrDefault()?.Name ?? PlaceholderSnippetField;
+            if (ExpansionSession.GetFieldSpan(firstField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            VsTextSpan adjustedTextSpan;
+            adjustedTextSpan.iStartLine = textSpan[0].iStartLine;
+            adjustedTextSpan.iStartIndex = textSpan[0].iStartIndex;
+
+            var lastField = _state._method?.Parameters.LastOrDefault()?.Name ?? PlaceholderSnippetField;
+            if (ExpansionSession.GetFieldSpan(lastField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            adjustedTextSpan.iEndLine = textSpan[0].iEndLine;
+            adjustedTextSpan.iEndIndex = textSpan[0].iEndIndex;
+
+            var snippet = CreateMethodCallSnippet(method.Name, includeMethod: false, method.Parameters, cancellationToken);
+            var doc = new DOMDocumentClass();
+            if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+            {
+                // Avoid clearing symbol information when InsertSpecificExpansion ends the current snippet session; the
+                // new session will need the same information to carry argument values forward.
+                _state._preserveSymbols = true;
+
+                _state._method = method;
+                var previousMethods = _state._methods;
+                var previousArguments = _state._arguments;
+
+                if (expansion.InsertSpecificExpansion(doc, adjustedTextSpan, this, LanguageServiceGuid, pszRelativePath: null, out _state._expansionSession) == VSConstants.S_OK)
+                {
+                    _state._preserveSymbols = false;
```

Comment:
So it's not clear to me why _preserveSymbols is necessary here -- you're already holding onto the local to assert the symbols aren't changing, why not just instead just reassign the _state? Then you can get rid of _preserveSymbols?

Reply:
The API is reentrant (other methods in this class run during a deeply-nested call stack) and eventually end up back here.

Reply:
So as best I can tell from the review is the reentrancy concern here is when you call insert specific the prior session will be dismissed, which if we don't set that wipes out state. But what other call backs are happening that need the state? Is it just the snippet functions which we think can be removed?